### PR TITLE
15286 Refactor to use allowed actions from API

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* bcgov/entity#15286 and bcgov/entity#15079

*Description of changes:*

Commit 1:
- app version = 5.12.0 (since this is a significant, but not major, change)
- imported latest shared Corp Type Module
- updated/added Allowable Actions mixin
- deleted obsolete hasBlockerTask/hasBlockerFiling code (components, store, etc)
- updated EntityInfo, Dashboard, etc to use updated isAllowed() method (including small tweaks to rendered vs disabled)
- updated isAllowed() to use data from Legal API instead of local business logic
- updated unit tests

Commit 2:
- replaced some getLegalName with getEntityName
- cleaned up EntityInfo title and sub-title
- renamed some IDs
- deleted some unneeded getter refs
- added some getLegalName fallback strings
- added getEntityName getter
- updated unit tests

Commit 3:
- moved EntityInfo tombstone data to new component
- added some getters to get rid of `@State` usage
- misc cleanup
- created/updated unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/EntityInfo.vue
Comment (severinbeauvais) [inline src/components/EntityInfo.vue:8]: Change from string to boolean.

File: src/components/EntityInfo.vue
Comment (severinbeauvais) [inline src/components/EntityInfo.vue:50]: I tweaked the logic a bit here to make more sense:
- don't show unless this is a business (not a temporary registration)
- don't show if business is historical
- disable if not allowed

Same below for Dissolve Business.

File: src/components/EntityInfo.vue
Comment (severinbeauvais) [inline src/components/EntityInfo.vue:167]: This is now always true, so I just made it a static class attribute.

Same below for Phone.

File: src/mixins/allowable-actions-mixin.ts
Comment (severinbeauvais) [inline src/mixins/allowable-actions-mixin.ts:34]: Dissolution type is not yet provided by the API - see ticket 15548. There is a note in that ticket to also patch this code and Voluntary Dissolution below, but the current code works as-is.

File: src/router.ts
Comment (severinbeauvais) [inline src/router.ts:60]: Just cleanup in here.

File: src/store/getters.ts
Comment (severinbeauvais) [inline src/store/getters.ts:58]: These 2 getters are no longer needed (including the "blocker task" and "blocker filing" states).

File: src/views/Dashboard.vue
Comment (severinbeauvais) [inline src/views/Dashboard.vue:51]: This check was useless -- it is always true if the current user is staff -- which is already checked on line 49.

File: src/components/Dashboard/StaffNotation.vue
Comment (severinbeauvais) [inline src/components/Dashboard/StaffNotation.vue:69]: This is cleanup and more concise (ie, driven by BE logic).

Basically, I display all items that are applicable for the current state (active or historical), and then disable the ones that aren't allowed.

File: src/components/Dashboard/StaffNotation.vue
Comment (severinbeauvais) [inline src/components/Dashboard/StaffNotation.vue:106]: This diff is easier to view if you hide whitespace changes.

File: src/components/Dashboard/StaffNotation.vue
Comment (severinbeauvais) [inline src/components/Dashboard/StaffNotation.vue:213]: Obsolete getters.

File: src/interfaces/business-interfaces.ts
Comment (severinbeauvais) [inline src/interfaces/business-interfaces.ts:8]: There are only 4 subtypes at the moment (all related to restorations). There may be more values for dissolution types. When we know more (future) then we can improve this.

File: package.json
Comment (severinbeauvais) [inline package.json:23]: Contains updated FilingNames enum (ie, for 4 restoration types).

File: src/App.vue
Comment (severinbeauvais) [inline src/App.vue:233]: Using function right from shared repo (not locally re-exported) -- as I removed it from Enum Mixin (which I hope to get rid of entirely soon).

File: src/App.vue
Comment (severinbeauvais) [inline src/App.vue:621]: Now using method from utilities class instead of Enum Mixin.

File: src/components/Dashboard/FilingHistoryList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/FilingHistoryList.vue:516]: Obsolete. This business logic is now computed in the Legal API.

File: src/components/Dashboard/TodoList.vue
Comment (severinbeauvais) [inline src/components/Dashboard/TodoList.vue:628]: Obsolete. This business logic is now computed in the Legal API.

File: src/components/EntityInfo.vue
Comment (severinbeauvais) [inline src/components/EntityInfo.vue:302]: I moved these here from the store getters file since this is the only place they are used.

File: src/enums/dissolutionTypesNames.ts
Comment (severinbeauvais) [inline src/enums/dissolutionTypesNames.ts:1]: This is now rolled into FilingSubTypes.

File: src/enums/restorationTypes.ts
Comment (severinbeauvais) [inline src/enums/restorationTypes.ts:1]: This is now rolled into FilingSubTypes.

File: src/interfaces/history-item-interface.ts
Comment (severinbeauvais) [inline src/interfaces/history-item-interface.ts:70]: This is a fix. It's never been a JS Date; always a string.

File: src/mixins/allowable-actions-mixin.ts
Comment (severinbeauvais) [inline src/mixins/allowable-actions-mixin.ts:123]: There is still code that uses this case. Ideally, in the future, code would look at only FullRestoration and LimitedRestoration cases.

This is actually even incorrect, since there are 2 other restoration cases that aren't included here. Oh well.

File: src/mixins/allowable-actions-mixin.ts
Comment (severinbeauvais) [inline src/mixins/allowable-actions-mixin.ts:127]: Eventually, this and maybe some other cases, will also come from the Legal API. For now this is fine.

File: src/mixins/enum-mixin.ts
Comment (severinbeauvais) [inline src/mixins/enum-mixin.ts:1]: Gradually making less and less use of this mixin so we can delete it.

File: src/services/enum-utilities.ts
Comment (severinbeauvais) [inline src/services/enum-utilities.ts:144]: I realized this method, and the ones below, are called from different components using different data objects. So, I addd the "param" comment indicating the items the method supports, and the code checks all possibilities.

I don't like this, but fixing it would require some API changes (since the API doesn't provide filing sub-type consistently) and maybe splitting this method into separate cases for API objects or UI objects. I think this is OK for now but I admit it's fragile.

File: src/services/enum-utilities.ts
Comment (severinbeauvais) [inline src/services/enum-utilities.ts:317]: I didn't move the dissolution subtype checks here because they check yet another parameter (isFirm), which THIS method doesn't know about. Sigh, the webs we weave.

File: tests/unit/enum-mixin.spec.ts
Comment (severinbeauvais) [inline tests/unit/enum-mixin.spec.ts:1]: Actually, I renamed this file (to enum-utilities) and changed what it tests.

File: src/components/EntityInfo.vue
Comment (jonathan-longe) [inline src/components/EntityInfo.vue:12]: What do you think of taking the opportunity to refactor the EntityInfo component?  It looks like we might be mixing a number of different display elements:
- entity name, 
- entity status, 
- entity header menu, and; 
- entity key facts

File: src/components/EntityInfo.vue
Comment (severinbeauvais) [reply src/components/EntityInfo.vue:302]: OK, I'll move it back.

File: src/components/EntityInfo.vue
Comment (jonathan-longe) [reply src/components/EntityInfo.vue:12]: A single component is fine until it either gets too big or too complicated.  In this case, I was thinking that splitting the component into separate parts makes sense because we use an `EntityInfo` component in the other repos but without the menu and the key facts.  By having a component that just displays the entity name and status, it would allow us to reuse the component in multiple places and have a consistent look as users jump from application to application.  (I'm stepping out on a limb here because I'm assuming the UX folks haven't asked from the `EntityInfo` header to vary between applications.)

File: src/mixins/enum-mixin.ts
Comment (jonathan-longe) [inline src/mixins/enum-mixin.ts:196]: Nice!  It's great to see these deprecated methods deleted.

File: src/services/enum-utilities.ts
Comment (jonathan-longe) [reply src/services/enum-utilities.ts:144]: Good work @severinbeauvais

File: src/App.vue
Comment (jonathan-longe) [reply src/App.vue:233]: What do you think of moving `this.getLegalName || GetCorpNumberedDescription(this.getLegalType)` to a store getter?  If I remember correctly, we reuse this logic in the EntityInfo component.   As we've seen with the sole proprietorship discussion, the logic behind what business name to display can get complicated.  I would suggest that any logic like this belongs in a single place where all components can access it.

File: src/components/EntityInfo.vue
Comment (severinbeauvais) [reply src/components/EntityInfo.vue:12]: Good points. I'm glad you're thinking clearly, haha.

I'll check on what blocks might be common between our 3 UIs, and if it makes sense then I'll create sub-components for them here. Later, we can reuse those sub-components in the other UIs.

File: src/App.vue
Comment (severinbeauvais) [reply src/App.vue:233]: Good idea, will do.

File: src/components/Dashboard/StaffNotation.vue
Comment (severinbeauvais) [inline src/components/Dashboard/StaffNotation.vue:100]: My thought process here is that this is never applicable unless it's a firm, so only show it for a firm (line 97) and then enable/disable it accordingly (line 100).

File: src/components/Dashboard/StaffNotation.vue
Comment (severinbeauvais) [inline src/components/Dashboard/StaffNotation.vue:1]: This diff is easier to view if you hide whitespace changes.

File: src/components/Dashboard/StaffNotation.vue
Comment (severinbeauvais) [inline src/components/Dashboard/StaffNotation.vue:111]: My thought process here is that this is never applicable unless the businss is historical.

**Review by JazzarKarim — APPROVED** [review]
LGTM! 👍

File: src/components/Dashboard/StaffNotation.vue
Comment (severinbeauvais) [inline src/components/Dashboard/StaffNotation.vue:151]: I don't know if this should be shown in all cases (eg, historical?) but we can change this easily in the future as we get UX feedback.

File: src/enums/allowableActions.ts
Comment (severinbeauvais) [inline src/enums/allowableActions.ts:23]: These don't all come from the Legal API yet, but most do 🎉

File: src/interfaces/api-task-interface.ts
Comment (severinbeauvais) [inline src/interfaces/api-task-interface.ts:57]: Added missing consentContinuationOut and sorted properties.

File: src/views/Dashboard.vue
Comment (severinbeauvais) [inline src/views/Dashboard.vue:120]: I moved the "is True" condition first and the "else" condition below. I think it's easier to read/understand this way: First, what does this apply to? Then, what's the fallback?

File: tests/unit/allowable-actions-mixin.spec.ts
Comment (severinbeauvais) [inline tests/unit/allowable-actions-mixin.spec.ts:22]: ^^ scoped helper functions

File: src/App.vue
Comment (severinbeauvais) [reply src/App.vue]: Updated in second commit. But there wasn't as much commonality as we thought.

File: src/components/Tombstone.vue
Comment (severinbeauvais) [inline src/components/Tombstone.vue:1]: This is a terrible name but I'm still not thinking clearly. Ideas?

File: src/components/EntityInfo.vue
Comment (severinbeauvais) [reply src/components/EntityInfo.vue]: I identified that only the right hand column ("tombstone") data might be common between Create/Edit/Filings UI, so I refactored that into its own sub-component. See third commit of this PR.

File: src/components/Tombstone.vue
Comment (JazzarKarim) [reply src/components/Tombstone.vue:1]: How about TombstoneProfile.vue? Do you reckon it's more specific in this case?

Comment (severinbeauvais) [reply src/components/Tombstone.vue:1]: It's more than just the profile info.

I was thinking TombstoneInfo, but that's long and redundant.

Keep up the ideas...

**Review by ketaki-deodhar — APPROVED** [review]

**Comment by severinbeauvais** [comment]
As this PR was built upon previous code (release 5.11.5), I will merge it to that branch. This will allow PR 467, which is built upon THIS PR, to be more easily compared for review purposes.

Still to do:
1. create a new PR for the content in THIS PR and merge it to main branch
2. rebase PR 467 to main

**Comment by severinbeauvais** [comment]
I created new PR https://github.com/bcgov/business-filings-ui/pull/478 that is rebased to the latest main branch and then contains the allowed actions changes. That PR is ready to merge once the BE changes are merged.

**Pull request merged**
